### ISSUE TITLE
Company internal CRAN-like repo 

### ISSUE THREAD

**Issue by fontbennz** [rstudio/renv#264]
Hi, 

From within an `renv` project, I have found that if I want to install from my companies internal CRAN-like repo (that requires authentication that is somehow managed automatically), I need to run:
`install.packages("custompkg", type = "source")`
If I don't specify `type = source`, it tries unsuccessfully to find a windows binary to install from (as they are not hosted).
However, if I am not in an `renv` project, I can run
`install.packages("custompkg")`
and it installs fine.

The real problems comes when creating a snapshot and then trying to restore.
This [issue](https://github.com/rstudio/renv/issues/120#issue-474858634) tells me why my package ends up as an "unknown" source.
I then edit the lock file to specify a repo, and then it tries to access the "Archive" location when running `renv::restore()`
```
Retrieving 'http://internalCRAN/src/contrib/Archive/custompkg/custompkg_1.5.1.tar.gz' ...
Error: failed to retrieve package 'custompkg'
In addition: Warning messages:
1: could not retrieve available packages for url "http://internalCRAN/src/contrib" 
2: could not retrieve available packages for url "http://internalCRAN/bin/windows/contrib/3.5" 
3: In downloader(url, destfile, type, request, headers) :
  curl: (22) The requested URL returned error: 404 Not Found
4: In downloader(url, destfile, type, request, headers) :
  curl: (22) The requested URL returned error: 404 Not Found
5: In downloader(url, destfile, type, request, headers) :
  curl: (22) The requested URL returned error: 401 Unauthorized
6: In downloader(url, destfile, type, request, headers) :
  curl: (22) The requested URL returned error: 401 Unauthorized
```
So, I created the Archive directory in our companies CRAN and placed the tar.gz files into that location, but I get the same error and warnings.

Do you have any ideas on how to make this work?

At the moment, I am resigned to using the "unknown" source method of copying the required tar.gz files and setting up a local directory where they can be installed from.
But I would like to be able to install directly from our internal CRAN-like repo directly from within a `renv::restore()`

Thanks,
Ben

**Comment by slopp** [comment]
It sounds like your company's CRAN-like mirror is not archiving the appropriate versions - are you sure when you manually created the Archive folder and moved the old package tar files there that  you copied the right structure and package version?

This is usually a sign that the CRAN-like repo isn't managed correctly. I am curious if you know how your company's repo is setup? We test renv with RStudio Package Manager (a tool for organizations to create their own CRAN-like repos) and haven't hit any errors. That might be an interesting option to save you and your IT group some pain.

**Comment by kevinushey** [comment]
The errors imply that `renv` (through `curl`) was denied access to your repository:

```
1: could not retrieve available packages for url "http://internalCRAN/src/contrib" 
2: could not retrieve available packages for url "http://internalCRAN/bin/windows/contrib/3.5" 
< ... >
5: In downloader(url, destfile, type, request, headers) :
  curl: (22) The requested URL returned error: 401 Unauthorized
6: In downloader(url, destfile, type, request, headers) :
  curl: (22) The requested URL returned error: 401 Unauthorized
```

Note that `renv` uses `curl` for downloads, which may not be the same as the default downloader set by R. What is the output of:

```
options("download.file.method", "download.file.extra")
```

on your machine?

You might also find this part of the introduction vignette useful: https://rstudio.github.io/renv/articles/renv.html#downloads

**Comment by fontbennz** [comment]
Hi Kevin,

The output of:
`options("download.file.method", "download.file.extra")`
is:

```
$download.file.method
[1] "wininet"

$download.file.extra
NULL
```

Thanks,
Ben

**Comment by fontbennz** [comment]
Hi Sean @slopp,

I was one of the R users that asked the IT team to provide a server.
It was up to myself and a colleague to set up the structure.

When we build packages, we use:
`miniCRAN::addLocalPackage()`
to write to the company CRAN server.
We just keep writing new tar.gz versions to the top level directory (no archive).
So I manually just created the "Archive" folder and any requisite package name folder before copying the files into the Archive (not removing them from the top level).

Thanks,
Ben

**Comment by fontbennz** [comment]
@slopp I'm considering using RStudio Package Manager and if it can help with properly managing our internal packages and then integrating nicely with renv, it might help to build a business case for Rstudio Package Manager.
We currently use Shiny Server Pro and are considering a switch to Rstudio Connect.

**Comment by kevinushey** [comment]
You could try setting the environment variable:

```
RENV_DOWNLOAD_FILE_METHOD = wininet
```

to force `renv` to use `wininet` rather than `curl` for downloads.

**Comment by fontbennz** [comment]
@kevinushey 
I added to my `.Renviron` file:
`RENV_DOWNLOAD_FILE_METHOD = wininet`
but I am still getting 401 unauthorized error.

```
5: In (function (url, destfile, method, quiet = FALSE, mode = "w",  :
  cannot open URL 'http://internalCRAN/src/contrib/Archive/cado/cado_0.7.2.9000.tar.gz': HTTP status was '401 Unauthorized'
```

**Comment by imanuelcostigan** [comment]
I have a very similar issue. In my case, the CRAN like repo is sitting on a network share drive (access controlled). In my case the URL looks something like this: `"file:///R:/drat"`. I use `drat` to manage package insertion. Here's my backtrace

```
> rlang::last_error()
<error/rlang_error>
package 'corppkg' is not available
Backtrace:
 1. renv:::install.packages("corppkg")
 2. renv::install(pkgs)
 3. renv:::renv_retrieve(packages)
 5. renv:::renv_retrieve_impl(package)
 6. renv:::renv_retrieve_repos(record)
 7. renv:::renv_available_packages_latest(record$Package)
 8. renv:::stopf("package '%s' is not available", package)
Call `rlang::last_trace()` to see the full backtrace.
```

**Comment by kevinushey** [comment]
@imanuelcostigan can you please file a separate issue with more details?

**Comment by kevinushey** [comment]
@fontbennz sorry for the trouble -- can you try updating to the latest development version of `renv` once more and let me know if you have any better luck?

**Comment by fontbennz** [comment]
@kevinushey I've updated `renv` to version 0.8.3.
I still get the same '401 Unauthorized' errors with both "curl" and "wininet".
Same error messages as above.

**Comment by fontbennz** [comment]
@kevinushey Ok, I installed version 0.8.3-55 and using with the "wininet" setting now works!
"curl" doesn't.

When will this version be made available in CRAN?

**Comment by kevinushey** [comment]
Great!

I'm planning to push another release of `renv` to CRAN soon -- hopefully in the next couple of weeks.

**Comment by fontbennz** [comment]
@kevinushey Thanks for your rapid response to this issue.
Looking forward to the next CRAN release that will have the resolution for this issue.

**Comment by fontbennz** [comment]
@kevinushey Some of my colleagues have been having problems restoring projects that worked for me after the update to `renv` version 0.9.2.
The error is the same as I found [here](https://github.com/rstudio/renv/issues/301#issuecomment-565115797)
For my colleague, when setting in the .Renviron file:
`RENV_DOWNLOAD_FILE_METHOD = wininet`
when attempting to download packages from our internal company CRAN-like repo, the error is:
`InternetOpenUrl failed: 'The requested header was not found'`
Also, it then fails to install packages from public CRAN with the following error:
```
Error: package 'remotes' is not available
In addition: Warning messages:
1: could not retrieve available packages for url "https://cran.csiro.au/src/contrib"
```

Then, if I remove `RENV_DOWNLOAD_FILE_METHOD = wininet` from the .Renviron file, then renv can install from public CRAN, but when attempting to install from the company internal CRAN-like repo, then the error is back to the authentication issue that led to using "wininet".

Is this something in `renv` or something to do with my colleagues Windows 10 machine and InternetOpenUrl?

**Comment by kevinushey** [comment]
Does this only occur when using `renv`? What happens if you try to run:

```
url <- "https://internalcran.com/src/contrib/PACKAGES.rds"  # replace with internal repo URL
file <- tempfile(fileext = ".rds")
download.file(url, file, method = "wininet")
```

Does that download the PACKAGES.rds file without issue?

Can you also try running the following:

```
trace(utils::download.file, quote(str(as.list(environment()))))
renv::restore()
```

and let me know what is printed out by the tracer when you do so?

**Comment by fontbennz** [comment]
Thanks, I'll need to wait till my colleague is back from Annual Leave, as it is on her laptop that the issue occurs, so please leave the issue open until I can respond again.

**Comment by fontbennz** [comment]
Hi, yes, the problem only occurs with renv. My colleague can install from our company internal CRAN before we started trying to use renv.

The next steps you asked me to try, I have then tried from within the renv activated project:
For the first test, the PACKAGES.rds file did not download.
```
In addition: Warning message:
In download.file(url, file, method = "wininet") :
  cannot open URL 'http://internalcran.com/src/contrib/PACKAGES.rds': HTTP status was '404 Not Found'
```
If I use a web browser and browse to `http://internalcran.com/src/contrib/`, I can see the PACKAGES.rds file. But if I click on PACKAGES.rds, instead of downloading the file, it takes me to a 404 not found page.

For the second test using TRACE and then `renv::restore()` on my colleagues machine, the outputs are:
```
Error: failed to retrieve package 'internalpackage'
In addition: Warning messages:
1: could not retrieve available packages for url "http://internalcran.com/src/contrib/src/contrib" 
2: could not retrieve available packages for url "http://internalcran.com/bin/windows/contrib/3.6" 
3: In system2("curl", args$data(), stdout = TRUE, stderr = TRUE) :
  running command '"curl" --config "C:/Users/user/AppData/Local/Temp/RtmpUdSJLY/renv-tempdir-f2901b2f53fd/renv-download-config-f29032d51b12"' had status 22
4: In downloader(url, destfile, type, request, headers) :
  curl: (22) The requested URL returned error: 404 Not Found
5: In system2("curl", args$data(), stdout = TRUE, stderr = TRUE) :
  running command '"curl" --config "C:/Users/user/AppData/Local/Temp/RtmpUdSJLY/renv-tempdir-f2901b2f53fd/renv-download-config-f290666a7fd4"' had status 22
6: In downloader(url, destfile, type, request, headers) :
  curl: (22) The requested URL returned error: 404 Not Found
7: In system2("curl", args$data(), stdout = TRUE, stderr = TRUE) :
  running command '"curl" --config "C:/Users/user/AppData/Local/Temp/RtmpUdSJLY/renv-tempdir-f2901b2f53fd/renv-download-config-f290e3a6654"' had status 22
8: In downloader(url, destfile, type, request, headers) :
  curl: (22) The requested URL returned error: 401 Unauthorized
9: In system2("curl", args$data(), stdout = TRUE, stderr = TRUE) :
  running command '"curl" --config "C:/Users/user/AppData/Local/Temp/RtmpUdSJLY/renv-tempdir-f2901b2f53fd/renv-download-config-f2901cee548a"' had status 22
10: In downloader(url, destfile, type, request, headers) :
  curl: (22) The requested URL returned error: 401 Unauthorized
```

**Comment by kevinushey** [comment]
> If I use a web browser and browse to http://internalcran.com/src/contrib/, I can see the PACKAGES.rds file. But if I click on PACKAGES.rds, instead of downloading the file, it takes me to a 404 not found page.

This sounds like it could be an issue with the repository itself. How is this package repository hosted? Are there some unique access restrictions set up? Is the machine and its version of Windows up-to-date?

I think you may need to resolve this issue before the `renv` issue can be resolved.

**Comment by fontbennz** [comment]
I can use renv to restore packages on my machine, despite the fact that I cannot use a web browser to download PACKAGES.rds
Our internal CRAN has an authentication system (I do not know what) that wininet seems to be able to pass through credentials just fine.
I can use the Windows file explorer to connect to the server and I can manually copy the PACKAGES.rds file onto my machine, and I can then load it into R to see details of the packages that exist in the repository.

My colleagues Windows 10 machine has an older OS build version than on my machine.
Once she is back, I'll ask her to see if she can update the OS build (may need our company IT team to run the update), to see if that fixes the issue.

**Comment by fontbennz** [comment]
After coming back from a break, our laptops received updates to the system from the IT team and my colleague and I are on the same Windows OS build version now.
This appears to have resolved the problem, as my colleague can now `renv::restore()` from our internal CRAN successfully.

**Comment by kevinushey** [comment]
Phew! Very glad to hear it.
